client/client.go

```diff
@@ -1199,14 +1199,17 @@ func (c *Client) syncConsul() {
 				for taskName, taskState := range taskStates {
 					if taskState.State == structs.TaskStateRunning {
 						if tr, ok := ar.tasks[taskName]; ok {
-							runningTasks = append(runningTasks, tr.task)
+							for _, service := range tr.task.Services {
+								services[service.ID(ar.alloc.ID, tr.task.Name)] = struct{}{}
+							}
 						}
 					}
 				}
 			}
-			if err := c.consulService.KeepServices(runningTasks); err != nil {
+			if err := c.consulService.KeepServices(services); err != nil {
 				c.logger.Printf("[DEBUG] client: error removing services from non-running tasks: %v", err)
 			}
+			sync = time.After(consulSyncInterval)
```

Comment:
Actually since you never cancel the thing you can use https://golang.org/pkg/time/#Tick